### ISSUE TITLE
Update Contribution guide with `package <pkg>_test` pattern for all testing

### ISSUE THREAD

**Issue by travisperson** [filecoin-project/go-filecoin#1995]
### Description

This idea comes from work @shannonwells has done in two PRs

- #1939 
- #1795

This is the first time I've seen this pattern of using `<pkg>_test`, pretty cool what it provides.

- It makes it trivial to run all the tests in a single file (if test files do not share common functionality defined elsewhere, this is great for running tests on file save.
- Help ensure we are writing testable code.

Here are some resource about this kind of testing

- https://stackoverflow.com/a/31443271
- https://medium.com/@matryer/5-simple-tips-and-tricks-for-writing-unit-tests-in-golang-619653f90742

One issue of using the different package name is we lose access to private intervals of packages. However, I would say in general being dependent on the intervals of a package is not a great way to test them. Public API testing is great, and using this approach will help us make sure our code is testable, and reduce the likely hood of breaking tests when implementations change. It's also nice to be able to make changes and have the assurance that the behavior didn't change. Tests that accidentally depend on intervals make this harder.

I think we could default to using `package <pkg>_test` for all `<file>_test.go` that are testing `<file>`. When we determine that we need internal access to properly test something (hopefully few and far between) we could use the pattern `package <pkg>`, and the name for the test file would be `<file>_internal_test.go`.

This will also be a nice flag to indicate that we are depending on package internals. Authors and reviewers can use this as feedback to look into if we could improve the code to make testing it no dependent on internals of the package.

To summarize
- All tests should default to using `<file>_test.go` with `package <pkg>_test` for all exports of `<file>`.
- When interval state information is required, tests should be placed in `<file>_interval_test.go` with `package <pkg>`.

Thanks @shannonwells, I think this is something we should look into.

**Comment by shannonwells** [comment]
Another thing it does is it forces one to think about code design. If one is reaching into unexported functions for your testing, then this *could* be a code smell.  It's better to test the package the way it's meant to be used, I think.

On the other hand, there is this "white box" vs. "black box" testing idea described in this [Stack Overflow thread](https://stackoverflow.com/questions/19998250/proper-package-naming-for-testing-with-the-go-language).  

Maybe it's worth considering *when* to make a test that's part of the package under test.  Perhaps if we do "white box" unit tests, then they should be in their own file and named `<package>_unit_test.go`, and its job is solely for unit tests that can't be done without direct access to private functions.

**Comment by phritz** [comment]
- agree that we should prefer foo_test for tests packages and that it enforces good practices. a small tradeoff is that you sometimes need to export more than you would otherwise which can muddy the public exported symbols but i dont really care about that too much.
- i dont think we need the extra process/complexity of having a convention around file naming that reflects this
- i think this suggestion is sufficiently non-controversial that you should just prep a change to https://github.com/filecoin-project/go-filecoin/blob/master/CONTRIBUTING.md#testing adding it and then make an announcement at the dev meeting

**Comment by rosalinekarr** [comment]
Renaming and relabelling per @phritz's suggestion to update `CONTRIBUTING.md`.
